core/src/main/java/com/linecorp/armeria/internal/common/metric/NettyMetricSupport.java

```diff
@@ -0,0 +1,125 @@
+ *  with the License. You may obtain a copy of the License at:
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *  License for the specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package com.linecorp.armeria.internal.common.metric;
+
+import com.linecorp.armeria.common.metric.MeterIdPrefix;
+import com.linecorp.armeria.common.util.Ticker;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.SingleThreadEventLoop;
+import io.netty.util.concurrent.EventExecutor;
+
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Registers stats to observe Netty's runtime.
+ *
+ * The following stats are currently exported:
+ *
+ *  - "num.workers" (gauge) - the total number of Netty's event loops
+ *  - "pending.tasks" (gauge) - the total number of IO tasks waiting to be run on event loops
+ */
+public class NettyMetricSupport {
+
+    static final long UPDATE_INTERVAL_NANOS = TimeUnit.SECONDS.toNanos(3);
+
+    public static void setup(MeterRegistry registry, MeterIdPrefix idPrefix, EventLoopGroup eventLoopGroup) {
+        requireNonNull(eventLoopGroup, "eventLoopGroup");
+
+        final NettyMetricSupport.NettyMetrics metrics = MicrometerUtil.register(
+                registry, idPrefix, NettyMetricSupport.NettyMetrics.class, NettyMetricSupport.NettyMetrics::new);
+
+        metrics.add(eventLoopGroup);
+    }
+
+    static final class NettyMetrics {
+
+        private Set<EventLoopGroup> registry = ConcurrentHashMap.newKeySet(2);
+
+        private final MeterRegistry parent;
+        private final MeterIdPrefix idPrefix;
+
+        private final Ticker ticker;
+
+        private volatile double lastPendingTasks = 0.0;
+        private volatile double lastNumWorkers = 0.0;
+        private volatile long lastStatsUpdateTime = 0L;
+
+        NettyMetrics(MeterRegistry parent, MeterIdPrefix idPrefix) {
+            this(parent, idPrefix, Ticker.systemTicker());
+        }
+
+        NettyMetrics(MeterRegistry parent, MeterIdPrefix idPrefix, Ticker ticker) {
+            this.parent = requireNonNull(parent, "parent");
+            this.idPrefix = requireNonNull(idPrefix, "idPrefix");
+            this.ticker = requireNonNull(ticker, "ticker");
+
+
+            String numWorkers = idPrefix.name("num.workers");
+            parent.gauge(numWorkers, idPrefix.tags(), this, metrics -> metrics.numWorkers());
+
+            String pendingTasks = idPrefix.name("pending.tasks");
+            parent.gauge(pendingTasks, idPrefix.tags(), this, metrics -> metrics.pendingTasks());
```

Comment:
The number of pending tasks for all event loops should be very useful. However, only some event loops can have pending tasks when:
- Performs a long-running task
- Waits for a lock
- Requests are unevenly distributed

It might be worth measuring `pending.tasks` for each event loop. What do you think?

Reply:
Yea I think @trustin was asking a similar question. Unless you're asking for a metric per worker thread? If so, I'd imagine it's douable but could be quite noisy given that there might lots of those. Also, I'm unsure what additional signal this finer granularity gives users in addition to "hey your IO runtime is slow and is probably under-provisioned". If, as a user, I see that `armeria-worker-1-5` is slow, how differently would I act on it compared to knowing that some/all armeria workers are slow?

Reply:
> I see that armeria-worker-1-5 is slow, how differently would I act on it compared to knowing that some/all armeria workers are slow?

Armeria client creates one connection for HTTP/2 by default. If the client heavily sends requests to the endpoint via the connection, the event loop allocated to the channel can be busy although other event loops are idle.

If all event loops are busy without blocking tasks, users may consider scaling up or out the servers. But if only some event loops are busy, it is possible to increase the number of event loops for the endpoint. It can distribute requests to multiple connections and event loops. 
https://github.com/line/armeria/blob/ec7c4251994f4b3b0089f95904dc1a68e5e9ce30/core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java#L182-L186

Reply:
> Armeria client creates one connection for HTTP/2 by default. If the client heavily sends requests to the endpoint via the connection, the event loop allocated to the channel can be busy although other event loops are idle.

Isn't there already a metric to get this signal (QPS per client)? Perhaps, even a higher-order one as we can pinpoint to a particular client.

Reply:
Perhaps we can do what @trustin was suggesting?

Two separate metric: one for ALL clients; one for ALL servers. I'm worried that anything beyond that would dilute the otherwise useful signal.

Reply:
> Isn't there already a metric to get this signal (QPS per client)? Perhaps, even a higher-order one as we can pinpoint to a particular client.

Right, it might be a way to monitor a client. Honestly, we got a report that only a few CPU cores have high loads and others do not. As event loops are CPU-intensive threads, I thought the metrics can be used to check if event loops cause the imbalance. That having said, it seems overkill in this PR. It would be better to add it later if there are any requests.

Reply:
:+1:

I will split this metric in two.

Reply:
@ikhoon mind taking another look this PR?